arenadata/db
# ADBDEV-4603: Allow multiple params in the child nodes of <dxl:TestExpr>

Allow multiple params in the child nodes of \<dxl:TestExpr> (#1140)

Problem definition:
When processing a query with a subquery (example in the test), the "Attribute
number ... not found in project list" error could be output. This happened in
the DXL to Plan Statement stage because the attribute was not found in the
list of column references (CMappingColIdVar) when processing the left node of
\<dxl:TestExpr>. \<dxl:TestExpr> is part of the \<dxl:SubPlan> node and the
TestExpr acts as a filter when a tuple is returned from SubPlan at execution
(see ExecScanSubPlan()).
\<dxl:TestExpr> is processed by TranslateDXLSubplanTestExprToScalar(). The
left node was processed depending on the flag "has_outer_refs". If it was true,
the TranslateDXLTestExprScalarIdentToExpr() was called to read one \<dxl:Ident>
and then convert it to a param. If it was false, the function of handling the
DXL tree was called without taking the param into account. In our case, the
"has_outer_refs" flag was false.
"has_outer_refs" was defined at the DXL generation stage in
PdxlnQuantifiedSubplan(). The params of TestExpr are the common columns of
the scalar and the inner part of PhysicalCorrelated***NLJoin node (see the
physical plan).
As it turns out, the left node can contain more than one param, and at this
stage we need to save the list of attributes that act as TestExpr params,
because in further processing we will not be able to determine them. So
instead of a flag "has_outer_refs", we need to form a list of TestExpr params.
However, if I fixed the determination of the TestExpr params, another error
occurred - "Unsupported subplan test expression", since the left node
contained a tree, but TranslateDXLSubplanTestExprToScalar() only supported
id and cast(id). Also, this function read the TestExpr param directly from
the \<dxl:Ident> node (without using CMappingColIdVar).
Therefore, the tree processing function (TranslateDXLToScalar) is used to
process the left node of TestExpr. In order for this function to correctly
process the TestExpr params, the corresponding list of column references
(CMappingColIdVar) is initialized.

Problem summary:
The \<dxl:TestExpr> handler could output errors:
"Attribute number ... not found in project list" and "Unsupported subplan
test expression".
The first error occurred if a reference to an attribute processed as a TestExpr
param (don't confuse with subplan params \<dxl:ParamList>) was not found. This
occurred due to incorrect detection of the presence of TestExpr params
(m_outer_param flag) at the DXL generation stage in the
PdxlnQuantifiedSubplan function and the lack of adding TestExpr params to the
list of references to attributes (CMappingColIdVar) at the \<dxl:TestExpr>
transforming stage.
The second error occurred if the left node \<dxl:TestExpr> contained a tree with
the TestExpr param, since only id and cast(id) were supported (the param was
read directly from the node).
See the minidump for more information, but keep in mind that the TestExpr param
presence flag (m_outer_param) is not serialized.
Support for one param in the left node of TestExpr was added in ac6ce1a.

Possible solutions:
The "Unsupported subplan test expression" error is fixed by using the
TranslateDXLToScalar function, which allows processing the tree in the child
nodes of TestExpr (mostly the left node). But in this case, it is impossible
to get the TestExpr param (currently it is read directly from node). It is
impossible to determine whether this is a param or a Var (there are no
markers indicating this), if there is more than one attribute in the
tree - this will cause the error "Attribute number ... not found in project
list". Therefore, the TestExpr params must be obtained in a separate structure
 and added in advance to the list of references to attributes
(CMappingColIdVar). Determinating the TestExpr params and adding them to the
CMappingColIdVar before handling DXL fixes the "Attribute number ... not
found in project list" error, but since the left node may contain a tree,
this will cause the "Unsupported subplan test expression" error. Therefore,
these errors must be fixed simultaneously.

Solution:
1) Determine the necessary columns as TestExpr params at the DXL generating
stage;
2) Add references to attributes processed as TestExpr params at the DXL to
Plan Statement stage (prefilling CMappingColIdVar);
2) Add support for processing a tree with params in the left node of TestExpr.

Changes:
At the DXL generation stage (in the PdxlnQuantifiedSubplan function),
the determination of the TestExpr params for PhysicalCorrelated***NLJoin
nodes has been unified. Instead of the "has_outer_refs" flag, the presence
of TestExpr params is now determined by the array of columns -
m_test_expr_params (passed via the CDXLScalarSubPlan class, not serialized).
At the DXL to Plan Statement stage in the subplan handler, a copy of the
mapping (list of references to columns) of the subplan is created to
handle \<dxl:TestExpr>, including the TestExpr params (obtained from
m_test_expr_params). We use the copy because we do not want the TestExpr
params to be mixed with the subplan params.
In the \<dxl:TestExpr> handler, the TranslateDXLToScalar function is
used instead of TranslateDXLTestExprScalarIdentToExpr for the left node,
which allows processing trees.
For the right node \<dxl:TestExpr>, TranslateDXLToScalar is also used for
unification purposes. The TranslateDXLTestExprScalarIdentToExpr function
has been removed as unnecessary.

Ticket: ADBDEV-4603
